eip-0014.md

```diff
@@ -0,0 +1,877 @@
+    val deltaSupplyLP  = supplyLP1 - supplyLP0
+    val deltaReservesX = reservesX1 - reservesX0
+    val deltaReservesY = reservesY1 - reservesY0
+
+    val validDepositing = {
+        val sharesUnlocked = min(
+            deltaReservesX.toBigInt * supplyLP0 / reservesX0,
+            deltaReservesY.toBigInt * supplyLP0 / reservesY0
+        )
+        deltaSupplyLP <= sharesUnlocked
+    }
+
+    val validRedemption = {
+        val _deltaSupplyLP = deltaSupplyLP.toBigInt
+        // note: _deltaSupplyLP, deltaReservesX and deltaReservesY are negative
+        deltaReservesX.toBigInt * supplyLP0 >= _deltaSupplyLP * reservesX0 && deltaReservesY.toBigInt * supplyLP0 >= _deltaSupplyLP * reservesY0
+    }
+
+    val validSwap =
+        if (deltaReservesX > 0)
+            reservesY0.toBigInt * deltaReservesX * feeNum0 >= -deltaReservesY * (reservesX0.toBigInt * FeeDenom + deltaReservesX * feeNum0)
+        else
+            reservesX0.toBigInt * deltaReservesY * feeNum0 >= -deltaReservesX * (reservesY0.toBigInt * FeeDenom + deltaReservesY * feeNum0)
+
+    val validAction =
+        if (deltaSupplyLP == 0)
+            validSwap
+        else
+            if (deltaReservesX > 0 && deltaReservesY > 0) validDepositing
+            else validRedemption
+
+    sigmaProp(
+        validSuccessorScript &&
+        preservedFeeConfig &&
+        preservedErgs &&
+        preservedPoolNFT &&
+        validLP &&
+        validPair &&
+        noMoreTokens &&
+        validAction
+    )
+}
+```
+
+#### Swap proxy-contract
+
+Swap contract ensures a swap is executed fairly from a user's perspective. The contract checks that:
+* Assets are swapped at actual price derived from pool reserves. `X_output = X_reserved * Y_input * fee_num / (Y_reserved * 1000 + Y_input * fee_num)`
+* Fair amount of DEX fee held in ERGs. `F = X_output * F_per_token`
+* A minimal amount of quote asset received as an output in order to prevent front-running attacks.
+
+Once published swap contracts are tracked and executed by ErgoDEX bots automatically. Until a swap is executed it can be cancelled by a user who created it by simply spending the swap UTXO.
+
+##### Contract parameters:
+Constant            | Type       | Description
+--------------------|------------|---------------
+Pk                  | SigmaProp  | User PublicKey
+FeeNum              | Long       | Pool fee numerator (must taken from pool params)
+QuoteId             | Coll[Byte] | Quote asset ID
+MinQuoteAmount      | Long       | Minimal amount of quote asset
+DexFeePerTokenNum   | Long       | Numerator of the DEX fee in nanoERGs per one unit of quote asset
+DexFeePerTokenDenom | Long       | Denominator of the DEX fee in nanoERGs per one unit of quote asset
+PoolNFT             | Coll[Byte] | ID of the pool NFT (Used as a reference to a concrete unique pool)
+
+```scala
+{
+    val FeeDenom = 1000
+
+    val poolIn = INPUTS(0)
+
+    val validTrade =
+        if (INPUTS.size == 2 && poolIn.tokens.size == 4) {
```

Comment:
We can remove the condition `poolIn.tokens.size == 4` as this is already enforced in pool box